### PR TITLE
GCS adapts to node table pub sub

### PR DESCRIPTION
**Pull request by ffbin**
## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/latest/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failure rates at https://ray-travis-tracker.herokuapp.com/.
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested (please justify below)

### CODE REVIEW

File: src/ray/gcs/gcs_client/service_based_accessor.cc [hunk 1]
```diff
@@ -567,6 +587,45 @@ Status ServiceBasedNodeInfoAccessor::AsyncSubscribeBatchHeartbeat(
...
+    // again. If the entry was in the cache and the node was deleted, check
+    // that this new notification is not an insertion.
+    if (!was_alive) {
+      RAY_CHECK(!is_alive)
+          << "Notification for addition of a node that was already removed:" << node_id;
+    }
+  }
+
+  // Add the notification to our cache.
+  RAY_LOG(INFO) << "Insertion/Deletion notification for node id = " << node_id
```
Comment (raulchen) [inline src/ray/gcs/gcs_client/service_based_accessor.cc:613]:
```suggestion
  RAY_LOG(INFO) << "Received notification for node id = " << node_id
```

File: src/ray/gcs/gcs_client/service_based_accessor.cc [hunk 2]
```diff
@@ -567,6 +587,45 @@ Status ServiceBasedNodeInfoAccessor::AsyncSubscribeBatchHeartbeat(
...
+                << ", IsAlive = " << is_alive;
+  node_cache_[node_id] = node_info;
+
+  // If the notification is new, call registered callback.
+  GcsNodeInfo &cache_data = node_cache_[node_id];
```
Comment (raulchen) [inline src/ray/gcs/gcs_client/service_based_accessor.cc:618]: nit, this line can be moved into the if block

File: src/ray/gcs/gcs_client/service_based_accessor.cc [hunk 3]
```diff
@@ -399,33 +399,53 @@ Status ServiceBasedNodeInfoAccessor::AsyncSubscribeToNodeChange(
...
+                        const Status &status,
+                        const std::vector<GcsNodeInfo> &node_info_list) {
+      for (auto &node_info : node_info_list) {
+        HandleNotification(node_info);
+      }
+      if (done) {
+        done(status);
+      }
+    };
+    RAY_CHECK_OK(AsyncGetAll(callback));
```
Comment (raulchen) [inline src/ray/gcs/gcs_client/service_based_accessor.cc:423]: Can you update the comment of `AsyncSubscribeToNodeChange` and indicate that the callback will be also be called for existing nodes?

**Review by raulchen — APPROVED** [review]
LGTM. only some minor comments.

File: src/ray/gcs/gcs_client/service_based_accessor.cc (hunk 3 above)
Comment (ffbin) [reply src/ray/gcs/gcs_client/service_based_accessor.cc:423]: fixed

File: src/ray/gcs/gcs_client/service_based_accessor.cc (hunk 2 above)
Comment (ffbin) [reply src/ray/gcs/gcs_client/service_based_accessor.cc]: fixed

File: src/ray/gcs/gcs_client/service_based_accessor.cc (hunk 1 above)
Comment (ffbin) [reply src/ray/gcs/gcs_client/service_based_accessor.cc]: fixed

File: src/ray/raylet/node_manager.cc [hunk 4]
```diff
@@ -517,7 +517,9 @@ void NodeManager::NodeAdded(const GcsNodeInfo &node_info) {
   if (node_id == self_node_id_) {
     // We got a notification for ourselves, so we are connected to the GCS now.
     // Save this NodeManager's resource information in the cluster resource map.
-    cluster_resource_map_[node_id] = initial_config_.resource_config;
+    if (0 == cluster_resource_map_.count(node_id)) {
```
Comment (raulchen) [inline src/ray/raylet/node_manager.cc:520]: why need this if?

**Review by raulchen — APPROVED** [review]
LGTM

File: src/ray/raylet/node_manager.cc (hunk 4 above)
Comment (ffbin) [reply src/ray/raylet/node_manager.cc]: This if is used to filter duplicate data, and i have moved it to the beginning of the function.

**Pull request merged**